src/adapters.ts

```diff
@@ -1644,31 +1838,123 @@ export function AdapterL1<TBase extends Constructor<TxSender>>(Base: TBase) {
         l2BridgeAddress,
         this._providerL2()
       );
+      if (!(await this.isProtocolVersionNew())) {
+        const calldata = l2Bridge.interface.decodeFunctionData(
+          'finalizeDeposit',
+          tx.data
+        );
+        const proof = await this._providerL2().getLogProof(
+          depositHash,
+          successL2ToL1LogIndex
+        );
+        if (!proof) {
+          throw new Error('Log proof not found!');
+        }
+        return await l1Bridge.claimFailedDeposit(
+          (await this._providerL2().getNetwork()).chainId as BigNumberish,
+          calldata['_l1Sender'],
+          calldata['_l1Token'],
+          calldata['_amount'],
+          depositHash,
+          receipt.l1BatchNumber!,
+          proof.id,
+          receipt.l1BatchTxIndex!,
+          proof.proof,
+          overrides ?? {}
+        );
+      } else {
+        const l1AR = IL1AssetRouter__factory.connect(
+          l1BridgeAddress,
+          this._signerL1()
+        );
+        const l2ARInterface = IAssetRouterBase__factory.createInterface();
+        const l1ntv = await this.getNativeTokenVaultL1();
+        try {
+          const calldata = l2Bridge.interface.decodeFunctionData(
+            'finalizeDeposit',
+            tx.data
+          );
 
-      const calldata = l2Bridge.interface.decodeFunctionData(
-        'finalizeDeposit',
-        tx.data
-      );
+          const proof = await this._providerL2().getLogProof(
+            depositHash,
+            successL2ToL1LogIndex
+          );
+          if (!proof) {
+            throw new Error('Log proof not found!');
+          }
 
-      const proof = await this._providerL2().getLogProof(
-        depositHash,
-        successL2ToL1LogIndex
-      );
-      if (!proof) {
-        throw new Error('Log proof not found!');
+          // todo: generally, providing token address is not required for the NTV
+          // this sdk does it, but we should support the case even when it is not given
+          const assetData = encodeNTVTransferData(
+            calldata['_amount'],
+            calldata['_l2Receiver'],
+            calldata['_l1Token']
+          );
+
+          const assetId = await l1ntv.assetId(calldata['_l1Token']);
+          if (assetId === ethers.ZeroHash) {
+            throw new Error(
+              `Token ${calldata['_l1Token']} not registered in NTV`
+            );
+          }
+
+          // todo: this SDK assumes that users used new encoding version,
+          // however in general it should be capable of working with hte old version as well
```

Comment:
It's not clear what is meant by "this SDK" and general message is hard to understand. I'd suggest to elaborate.

Reply:
rewrote